### ISSUE TITLE
Add structural way to define background context within components

### ISSUE THREAD

**Issue by bheston** [microsoft/fast#2211]
The way the color recipes are currently applied works well in most simple components, but is more difficult to implement or customize in multipart components. This particularly affects components with floating elements and children.

- We have layer recipes for a semantic way to define layout that works in either light or dark mode, like `neutralLayerFloating`.
- One of the easiest ways to apply a background color for a component and have its children respond correctly is to use a `Background` component.
- Alternatively you need to override background and foreground colors for the container and all child components if some element is logically a different background color.
- For many component there is no way to inject `Background` functionality.

A great example of this problem is on the Select component. With current styling, the trigger button portion appears as if it's on the containing background, but the popup menu is floating. With the current color recipes, the menu should have a background of `neutralLayerFloating`. Applying this in the Select style works for the base menu element, but none of the SelectOption children render in the correct context.

One solution is to restyle the children such that all color recipes pass `neutralLayerFloating` as the context. This is error prone and doesn't encourage reuse and assembly.
An alternate solution would be to add render props for every piece of a component we might want to override. For instance, Select already exposes a `trigger` prop to change the button used to activate the menu. This might be a good approach, but I haven't discovered a way to simply wrap the underlying default implementation. That's all that's necessary in this situation.

The `Background` component and color recipes are contained in the `msft` component implementation, but the base components don't currently expose enough configuration points to make them easy to apply.

A similar need to inject or wrap something with a `Background` component has come up before and the solution we used at the time was to plumb it through as another prop where the color value could be controlled by the consumer. That was for component built on FAST msft styling though. We essentially need the equivalent in the base components.

**Comment by bheston** [comment]
Looking for ideas or conversation about how to solve this so I can make some related component styling updates.

**Comment by janechu** [comment]
> One solution is to restyle the children such that all color recipes pass neutralLayerFloating as the context. This is error prone and doesn't encourage reuse and assembly.

I'm not sure I'm understanding the issue with the component redefining the context, could you expand a little?
Are you saying there would be a problem with `@microsoft/fast-components-react-msft` Select creating its own context provider and redefining the `backgroundColor` with `neutralLayerFloating`? It seems like that would work.

**Comment by connorjsmith** [comment]
> An alternate solution would be to add render props for every piece of a component we might want to override. For instance, Select already exposes a trigger prop to change the button used to activate the menu. This might be a good approach, but I haven't discovered a way to simply wrap the underlying default implementation. That's all that's necessary in this situation.

How about something like this?
```typescript

function renderMenu(props: SelectProps, state: SelectState, defaultMenu: React.ReactNode): React.ReactNode {
  return (
  <Background value={...}>
    {defaultMenu}
  </Background>);
}

<Select
  menu={renderMenu}
/>
```

Looking at the use cases:
1. **Completely replace the menu (ie. no ListBox):** Rebuild from props and state, same as `trigger` is today
2. **Only wrap the menu:** Use example above
3. **Only style the menu:** jssStylesheet

Are there any cases I'm missing here? I feel like this pattern could be extended to all render functions, and wouldn't even constitute a breaking change if we make `children` the last item passed to a render prop.

**Comment by bheston** [comment]
Yeah, I was thinking something along those lines, at least for this particular example. There might still be a more general way to handle this structure for other components as well. I'll open a new issue for select specifically.

**Comment by chrisdholt** [comment]
closed with #3517 as the background component does not exist
